core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -560,11 +560,15 @@ protected RexNode removeCorrelationExpr(
       // Now add the corVars from the input, starting from
       // position oldGroupKeyCount.
       for (Map.Entry<CorDef, Integer> entry : frame.corDefOutputs.entrySet()) {
-        RexInputRef.add2(projects, entry.getValue(), newInputOutput);
-
-        corDefOutputs.put(entry.getKey(), newPos);
-        mapNewInputToProjOutputs.put(entry.getValue(), newPos);
-        newPos++;
+        final Integer pos = mapNewInputToProjOutputs.get(entry.getValue());
```

Comment:
I think a comment would help maintainers understand why this can happen

Reply:
Agree. Comment added.